signer/src/bitcoin/utxo.rs

```diff
@@ -1134,115 +1126,57 @@ impl<'a> UnsignedTransaction<'a> {
+    /// data (it is not constrained by the Bitcoin protocol itself but rather by
+    /// the miners' policy). We use this space to encode the withdrawal request
+    /// IDs.
     ///
-    /// A more full description of the merkle root can be found in the
-    /// comments for [`UnsignedTransaction::withdrawal_merkle_root`]. If
-    /// there are no withdrawal requests in the input slice then the layout
-    /// looks like:
+    /// ## Wire Format
+    /// The layout of the OP_RETURN output is as follows:
     ///
     /// ```text
-    ///  0       2    3     5        21
-    ///  |-------|----|-----|--------|
-    ///    magic   op   N_d   bitmap
+    ///  0       2    3                                           X<80
+    ///  |-------|----|--------------------------------------------|
+    ///    magic   op   [encoded withdrawal IDs (variable-length)]
     /// ```
     ///
-    /// In the above layout, magic is the UTF-8 encoded string "ST", op is
-    /// the version byte, and N_d is the of deposits in this transaction
-    /// encoded as a big-endian two byte unsigned integer.
+    /// In the above layout:
+    /// - magic: UTF-8 encoded string indicator (2 bytes)
+    /// - op: version byte (1 byte)
+    /// - encoded IDs: withdrawal request IDs encoded using idpack (variable
+    ///   length, if there are withdrawals serviced by the transaction)
     ///
-    /// `None` is only returned if there are no requests in the input slice.
-    fn new_op_return_output(reqs: &Requests, state: &SignerBtcState) -> Option<TxOut> {
-        let sbtc_data = Self::sbtc_data(reqs, state)?;
-        let script_pubkey = match Self::withdrawal_merkle_root(reqs) {
-            Some(merkle_root) => {
-                let mut data: [u8; 41] = [0; 41];
-                data[..21].copy_from_slice(&sbtc_data);
-                data[21..].copy_from_slice(&merkle_root);
-
-                ScriptBuf::new_op_return(data)
-            }
-            None => ScriptBuf::new_op_return(sbtc_data),
-        };
+    /// ## Returns
+    /// - `Some(TxOut)`: the resulting OP_RETURN output
+    fn new_op_return_output(reqs: &Requests, state: &SignerBtcState) -> Result<TxOut, Error> {
+        // Create OP_RETURN data
+        let mut data = PushBytesBuf::with_capacity(OP_RETURN_MAX_SIZE);
+        data.extend_from_slice(&state.magic_bytes)?;
+        data.push(OP_RETURN_VERSION)?;
+
+        // Extract all withdrawal request IDs
+        let withdrawal_ids: Vec<u64> = reqs
+            .iter()
+            .filter_map(|req| req.as_withdrawal().map(|w| w.request_id))
+            .collect();
 
-        let value = Amount::ZERO;
-        Some(TxOut { value, script_pubkey })
-    }
+        // If there are any withdrawal ID's, encode them and add them to the
+        // OP_RETURN data.
+        if !withdrawal_ids.is_empty() {
+            let encoded = BitmapSegmenter.package(&withdrawal_ids)?.encode();
+            data.extend_from_slice(&encoded)?;
```

Comment:
Do we know whether the encoded bytes are within `OP_RETURN`'s limits? I didn't think that the packager checked that but I'm not sure. We should probably check here anyway.

Reply:
Yeah, each bag keeps track of its current withdrawal id's and runs estimations including new items' id's as part of the best-fit algorithm, with each bag capped at `OP_RETURN_AVAILABLE_SIZE` (80 - 3 = 77 bytes).

But I agree, doesn't hurt with an extra check here, added in [c0fe873](https://github.com/stacks-network/sbtc/pull/1512/commits/c0fe8731db74dcb54a90e4f3c5c5232a4a1989e2)